src/Suggestor/Suggestor.ts

```diff
@@ -386,6 +427,50 @@ function isAnyBracketOpen(line: string, brackets: [opening_bracket: string, clos
     return Object.values(numOpeningBrackets).some((n) => n > 0);
 }
 
+/**
+ * Checks whether _any_ of the bracket pairs in {@link brackets} is open at the end of the string {@link line}
+ * If there are any open brackets, returns the last one. Else, returns null.
+ *
+ * @example
+ *     lastOpenBracket("(hello world",   [['(', ')']]);             // '('
+ *     lastOpenBracket("[hello world",   [['[', ']']]);             // '['
+ *     lastOpenBracket("[hello world",   [['(', ')'], ['[', ']']]); // '['
+ *     lastOpenBracket("([hello world)", [['(', ')'], ['[', ']']])  // '['
+ *     lastOpenBracket("))))(",          [['(', ')']])              // '('
+ *     lastOpenBracket("(hello world)",  [['(', ')']]);             // null
+ *     lastOpenBracket("(hello world)",  []);                       // null
+ *
+ * @param line - The line of text to scan
+ * @param brackets - A listed of tuples that defines bracket pairs.
+ * @returns The last open bracket in line among the given bracket pairs. If no such bracket exists, return null.
+ */
+function lastOpenBracket(line: string, brackets: [opening_bracket: string, closing_bracket: string][]): string | null {
+    if (brackets.length === 0) {
+        return null;
+    }
+    const numOpeningBrackets = Object.fromEntries(brackets.map(([open, _]) => [open, 0]));
+    const openingOf = Object.fromEntries(brackets.map(([open, close]) => [close, open]));
+    const openBracketsStack = [];
+    for (let i = 0; i < line.length; i++) {
+        const c = line[i];
+        if (c in numOpeningBrackets) {
+            numOpeningBrackets[c]++;
+            openBracketsStack.push({ bracket: c, idx: i });
+        } else if (c in openingOf) {
+            if (numOpeningBrackets[openingOf[c]] >= 1) {
+                for (let idx = openBracketsStack.length - 1; idx >= 0; idx--) {
+                    if (openBracketsStack[idx].bracket == openingOf[c]) {
+                        openBracketsStack.splice(idx, 1);
+                        break;
+                    }
+                }
+            }
+            numOpeningBrackets[openingOf[c]] = Math.max(0, numOpeningBrackets[openingOf[c]] - 1);
+        }
+    }
```

Comment:
Please could that logic be protected with unit tests? It will be easy for a future maintainer to make a mistake, and manually testing against the examples in the comments is not practical.

Reply:
I already add unit test and update the doc of Auto-Suggest.md.